File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.go
func productExceptSelf(nums []int) []int {
    l := make([]int, len(nums))
    r := make([]int, len(nums))
    res := make([]int, len(nums))

    l[0] = 1
    r[len(nums)-1] = 1

    for i := 1 ; i < len(nums) ; i++ {
        l[i] = l[i-1] * nums[i-1]
    
    }
    res[len(nums)-1] = l[len(nums)-1] * r[len(nums)-1]
    for i := len(nums)-2 ; i >= 0 ; i-- {
        r[i] = r[i+1] * nums[i+1]
        res[i] = r[i] * l[i]
    }


    return res
} 

File: Data Structures & Algorithms/maximum-product-subarray/submission-0.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)

        max_dp = [0] * n
        min_dp = [0] * n

        max_dp[0] = nums[0]
        min_dp[0] = nums[0]

        res = nums[0]

        for i in range(1, n):
            x = nums[i]

            max_dp[i] = max(
                x,
                max_dp[i-1] * x,
                min_dp[i-1] * x,
            )
            min_dp[i] = min(
                x,
                max_dp[i-1] * x,
                min_dp[i-1] * x,
            )

            res = max(res, max_dp[i])
        
        return res